File: python/81-90/Partition List.py
   # @param head, a ListNode
    # @param x, an integer
    # @return a ListNode
    def partition(self, head, x):
        if head==None or head.next==None:
            return head
        thead = ListNode(0)
        thead.next=head
        pre,cur=thead,head
        while cur!=None and cur.val<x:
            cur = cur.next
            pre = pre.next
        if cur==None:
            return head
        pt,t = pre,cur
        cur=cur.next
        pre=pre.next
        while cur!=None:
            if cur.val<x:
                pt.next=cur
                pt=pt.next
                pre.next=cur.next
                cur.next=t
                cur=pre.next
            else:
                cur = cur.next
                pre = pre.next
        return thead.next

File: Cpp/81-90/Partition List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
 
 /**
  * runtime : 12ms
  * error : 没有考虑else中pre与temp相等的情况！！ 记住 ！
  */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode thead(0);
        thead.next = head;
        ListNode* temp = &thead,*pre =&thead;
        while(temp->next!=nullptr){
            if(temp->next->val>=x)
                temp = temp->next;
            else{
                ListNode* t = temp->next;
                temp->next =temp->next->next;
                t->next = pre->next;
                pre->next = t;
                pre = t;
            }
        }
        return thead.next;
    }

/**
 *  another simple code!  great!
 *  runtime : 8ms
 *  
 * 
 */ 
class Solution1 {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode h1(0),h2(0);
        ListNode* l1 = &h1,*l2 =&h2;
        while(head!=nullptr){
            if(head->val>=x)   l2=l2->next=head;
            else   l1=l1->next=head;
            head=head->next;
        }
        l2->next = NULL;
        l1->next = h2.next;
        return h1.next;
    }
};
